File: dospy/extract.cc
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <iconv.h>

const int MAX_PAGE_LENGTH = 1048576; // 1M per page
const int MAX_FILE_NAME = 128;

int printUsage() {
  printf("Usage:\n");
  printf("  exDospy filename\n");
  return 0;
}

int gbk2utf8(char* in_str, char* out_str) {
  iconv_t cd = iconv_open("utf-8", "gbk");
  if (cd == 0)
    return -1;

  size_t instrlen = strlen(in_str);
  size_t outstrlen = MAX_PAGE_LENGTH;
  if (iconv(cd, &in_str, &instrlen, &out_str, &outstrlen) == -1)
    return -2;
  
  iconv_close(cd);
  return 0;
}

int grepQuote(char* str) {
  char *tp;
  char *rp;
  while (tp = strstr(str, "<div class=\"msgbody\"><div class=\"msgheader\">")) {
    rp = strstr(tp, "</div></div>") + 12;
    if (*rp == '\0') {
      *tp = '\0';
    } else {
      strncpy(tp, rp, strlen(rp) + 1);
    }
  }

  return 0;
}

int grepHtmlTag(char* str) {
  char *tp;
  char *rp;

  while (tp = strstr(str, "<")) {
    if (tp >= str + strlen(str))
      break;

    rp = strstr(tp, ">") + 1;

    if (*rp == '\0') {
      *tp = '\0';
    } else {
      strncpy(tp, rp, strlen(rp) + 1);
    }
  }

  return 0;
}

int convertHtmlChar(char* str) {
  char html_code[5][2][8] = {
    "&nbsp;", " ",
    "&amp;", "&",
    "&lt;", "<",
    "&gt;", ">",
    "&quot;", "\""
  };
  char* tp;

  for (int i = 0; i < 5; ++i) {
    int len = strlen(html_code[i][0]) - 1;
    while (tp = strstr(str, html_code[i][0])) {
      strncpy(tp, tp + len, strlen(tp + len));
      *tp = html_code[i][1][0];
    }
  }

  return 0;
}

int main(int argc, char* argv[]) {
  char target_file_name[MAX_FILE_NAME + 1];
  char buf[MAX_PAGE_LENGTH + 1];
  char page_content[MAX_PAGE_LENGTH + 1];
  char title[MAX_PAGE_LENGTH + 1];

  if (argc < 2 ) {
    printUsage();
    exit(0);
  } else if (argc < 3) {
    time_t t = time(NULL);
    struct tm tm;
    localtime_r(&t, &tm);
    strftime(target_file_name, MAX_FILE_NAME, "%Y%m%d-%H%M%S", &tm);
  } else {
    strncpy(target_file_name, argv[2], MAX_FILE_NAME);
  }

  int offset = strlen(target_file_name) + 1;
  strcat(target_file_name, ".0000");

  FILE* origin_file = fopen(argv[1], "r");
  memset(page_content, 0, sizeof(page_content));
  memset(buf, 0, sizeof(buf));

  while (!feof(origin_file)) {
    fgets(buf, MAX_PAGE_LENGTH, origin_file);
    strncat(page_content, buf, MAX_PAGE_LENGTH);
  }
  gbk2utf8(page_content, buf);
  strncpy(page_content, buf, MAX_PAGE_LENGTH);

  char* sp;
  char* tp;

  // detect title
  sp = strstr(page_content, "<title>");
  sp += 7;
  tp = strstr(sp, "-");
  strncpy(title, sp, tp - sp);
  title[tp - sp - 1] = '\0';

  FILE* target_file;
  int target_file_no = 0;

  while (sp = strstr(sp, "<div style=\"padding-top: 4px;\">")) {
    target_file_no++;
    sprintf(&target_file_name[offset], "%04d", target_file_no);
    target_file = fopen(target_file_name, "w");
    char date_time[MAX_PAGE_LENGTH + 1];
    char content[MAX_PAGE_LENGTH + 1];
    
    // detect date time info
    sp += 42;
    tp = strstr(sp, "&nbsp");
    strncpy(buf, sp, tp - sp);
    buf[tp - sp] = '\0';
    strncpy(date_time, buf, strlen(buf) + 1);

    // detect text
    sp = strstr(sp, "class=\"t_msgfont\">");
    sp += 18;
    tp = strstr(sp, "</div>\r\n");
    memset(buf, 0, sizeof(buf));
    strncpy(buf, sp, tp - sp);
    buf[tp - sp] = '\0';

    grepQuote(buf);
    grepHtmlTag(buf);
    convertHtmlChar(buf);
    strncpy(content, buf, strlen(buf) + 1);

    fprintf(target_file, "\n<DOC>\n");
    fprintf(target_file, "<DOCNO> %s </DOCNO>\n", target_file_name);
    fprintf(target_file, "<DATE_TIME> %s </DATE_TIME>\n", date_time);
    fprintf(target_file, "<BODY>\n");
    fprintf(target_file, "<CATEGORY> cellphone </CATEGORY>\n");
    fprintf(target_file, "<HEADLINE> %s </HEADLINE>\n", title);
    fprintf(target_file, "<TEXT>\n");
    fprintf(target_file, "%s\n", content);
    fprintf(target_file, "</TEXT>\n");
    fprintf(target_file, "</BODY>\n");
    fprintf(target_file, "</DOC>\n");
    fclose(target_file);
  }

  return 0;
}

File: dospy/extract.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Author: cswenye@gmail.com                                
#

__author__ = 'cswenye@gmail.com (Wen YE)'

import sys
import time
import optparse

import logging

def extractPage(pagefile, target = time.strftime('%Y%m%d-%H%M%S'), start_no = 0) :
  log.info('Extract %s to %s from %d' % (pagefile, target, start_no))

  exfilelist = []

  pf = file(pagefile, "r")
  page_content = pf.read()
  pf.close()

  # If ask to output in a single file
  if start_no == -1 :
    target_file_name = target
    tf = file(target_file_name, "w")

  # detect title
  title = detectTitle(page_content)

  # detect each post
  time_start = page_content.find('<div style="padding-top: 4px;">')
  while time_start != -1 :
    if start_no >= 0 :
      target_file_name = target + ".%04d" % start_no
      start_no += 1
      tf = file(target_file_name, "w")

    time_start += 42
    time_end = page_content.find('&nbsp;', time_start)
    date_time = page_content[time_start:time_end]

    content_start = page_content.find('class="t_msgfont">', time_start) + 18
    content_end = page_content.find('</div>\r\n', content_start)
    content = page_content[content_start:content_end]

    reply_start = page_content.find('<span class="bold">回复 #', time_start, content_start)
    log.info('find reply_start from %d to %d at %d: %s' % \
             (time_start, content_start, reply_start, page_content[reply_start+27:reply_start+40]))
    if reply_start == -1 :
      reply_id = -1
    else :
      reply_start += 27
      reply_id = int(page_content[reply_start:page_content.find(' ', reply_start)])
    
    content = grepLastEdit(content)
    content = grepAttach(content)
    content, quote = grepQuote(content)
    content = grepHtmlTag(content)
    content = convertHtmlChar(content)
    content = content.replace('\r', '')

    outcontent = "\n<DOC>\n" \
               + "<DOCNO> %s </DOCNO>\n" % target_file_name \
               + "<DATE_TIME> %s </DATE_TIME>\n" % date_time \
               + "<BODY>\n" \
               + "<CATEGORY> cellphone </CATEGORY>\n" \
               + "<HEADLINE> %s </HEADLINE>\n" % title \
               + "<REPLY> %d </REPLY>\n" % reply_id \
               + "<QUOTE> %s </QUOTE>\n" % quote \
               + "<TEXT>\n" \
               + "%s\n" % content \
               + "</TEXT>\n" \
               + "</BODY>\n" \
               + "</DOC>\n"
    tf.write(outcontent)

    if start_no >= 0 :
      tf.close()
      exfilelist.append(target_file_name)
  
    time_start = page_content.find('<div style="padding-top: 4px;">', time_start)

  if start_no == -1 :
    tf.close()
    exfilelist.append(target_file_name)

  return exfilelist

def detectTitle(content) :
  tsp = content.find('<div class="subtable nav"')
  ttp = content.find('</div>', tsp)
  while content.find('&raquo;', tsp, ttp) != -1 :
    tsp = content.find('&raquo;', tsp, ttp) + 7
  return content[tsp:ttp]

def grepLastEdit(content) :
  while True :
    tp = content.find('[<i> 本帖最后由')
    if tp == -1 :
      break
    rp = content.find('</i>]', tp) + 5
    content = content[0:tp] + content[rp:]
  return content

def grepAttach(content) :
  while True :
    tp = content.find('<span style="white-space:nowrap" id="attach_')
    if tp == -1 :
      break
    rp = content.find('</span>', tp) + 7
    content = content[0:tp] + content[rp:]

  while True :
    tp = content.find('<div title="menu" class="t_attach"')
    if tp == -1 :
      break
    rp = content.find('</div></div>', tp) + 12
    content = content[0:tp] + content[rp:]

  return content

def grepQuote(content) :
  quote = ""
  while True :
    tp = content.find('<div class="msgbody"><div class="msgheader">')
    if tp == -1 :
      break
    qs = content.find('\r\n') + 2
    qt = content.find('</div></div>', tp)
    rp = qt + 12
    quote = content[qs:qt]
    content = content[0:tp] + content[rp:]
  return content, quote

def grepHtmlTag(content) :
  while True :
    tp = content.find('<')
    if tp == -1 :
      break
    rp = content.find('>', tp) + 1
#    print '\n>>>>\n in %s grep %s \n>>>> after <<<<\n %s\n<<<<\n' % (content, content[tp:rp], content[0:tp] + content[rp:])
    content = content[0:tp] + content[rp:]
  return content

def convertHtmlChar(content) :
  html_tags = { '&nbsp;' : ' ',
                '&amp;'  : '&',
                '&lt;'   : '<',
                '&gt;'   : '>',
                '&quot;' : '"'
              }
  for html_tag, html_char in html_tags.items() :
#    print '\n>>>> before replace %s as %s >>>>\n%s' % (html_tag, html_char, content)
    content = content.replace(html_tag, html_char)
#    print '\n<<<< after replace %s as %s <<<<\n%s' % (html_tag, html_char, content)

  return content

def main() :
  global log
  log = logging.getLogger(__name__)
  log.setLevel(logging.DEBUG)
  formatter = logging.Formatter('%(asctime)s [%(levelname)s] %(filename)s:%(lineno)d: %(message)s', '%y-%m-%d,%H:%M:%S')
  ch = logging.StreamHandler()
  ch.setFormatter(formatter)
  log.addHandler(ch)

  parser = optparse.OptionParser(usage='%prog [options] FILE')
  parser.add_option('-o', '--output', dest='output',
                    help='Output filename, or use current time defaultly')
  parser.add_option('-s', '--single', dest='single', action='store_true', default=True,
                    help='Output file in a single file or a series files')
  
  options, args = parser.parse_args()

  start_no = options.single and -1 or 0

  if len(args) < 1:
    parser.error('File to extract not provided.')
  elif len(args) > 1:
    parser.error('Only one file may be specified.')

  file_path = args[0]

  if options.output:
    extractPage(file_path, options.output, start_no)
  else:
    extractPage(file_path, start_no=start_no)

  return 0

if __name__ == '__main__' :
  sys.exit(main())

